### ISSUE TITLE
The new ActionController::Parameters does not work well with STI (as it used to)

### ISSUE THREAD

**Issue by claudiob** [rails/rails#21407]
Hello :ribbon: 

I found that the change in the inheritance of  ActionController::Parameters from Hash breaks the use-case of creating a new instance of an ActiveRecord, whose subclass is specified by the user.

[I created a gist which fails on master](https://gist.github.com/claudiob/1257599ba9c704e9260d) but works the current Rails 4.2.3 is used instead.

---

Here is the use-case I'm talking about.

An ActiveRecord `Pill` class exists, with two subclasses `BluePill` and `RedPill`, created with STI (that is, with a string field `type` in the `pills` table).

A user browses to `/pills/new` to create a new pill and has a radio button to select which type of pill to build. 

When the user submits the form, the `params` either include `{"type" => "BluePill"}` or `{"type" => "RedPill"}`

This `type` param is expected, so it is whitelisted in the controller.

Therefore, the `create` action in the controller works as follows:

```ruby
def create
  @pill = Pill.new params.require(:pill).permit(:type)
  if @pill.save
    redirect_to @pill, notice: 'You have created a new pill'
  else
    render :new
  end
end
```

The code above works correctly in Rails 4.2.3. 
The `type` provided through the form is correctly passed to ActiveRecord.
If the user submits `{"pill" => {"type" => "BluePill"}}`, then `@pill` is correctly an instance of `BluePill`.

However, the code above does not work on the latest master.
As you can see running the gist, `@pill` is now instead an instance of `Pill`.

**Comment by claudiob** [comment]
If anyone is experiencing the same issue, I'd like to add that adding `.to_h` to the params makes the issue go away:

```ruby
@pill = Pill.new params.require(:pill).permit(:type).to_h
```
 However, I still wonder if this is an intended consequence of #14384 or something that needs to be changed. @sikachu ?

**Comment by matthewd** [comment]
That's a bug.

**Comment by claudiob** [comment]
@matthewd I investigated, and here's the reason why is the `is_a?(Hash)` check from [active_record/inheritance.rb#L195](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/inheritance.rb#L195):

```ruby
# Detect the subclass from the inheritance column of attrs. If the inheritance column value
# is not self or a valid subclass, raises ActiveRecord::SubclassNotFound
# If this is a StrongParameters hash, and access to inheritance_column is not permitted,
# this will ignore the inheritance column and return nil
def subclass_from_attributes?(attrs)
  attribute_names.include?(inheritance_column) && attrs.is_a?(Hash)
end
```

**Comment by sikachu** [comment]
Closing this since #21452 was closed, and the changes that required to fixed it was merged in #17169.
